instrumentation-genai/opentelemetry-instrumentation-agentscope/src/opentelemetry/instrumentation/agentscope/_wrapper.py

```diff
@@ -0,0 +1,130 @@
+from opentelemetry.instrumentation.agentscope._request_attributes_extractor import RequestAttributesExtractor
+from opentelemetry.instrumentation.agentscope._response_attributes_exatractor import ResponseAttributesExtractor
+from opentelemetry.instrumentation.agentscope._with_span import _WithSpan
+from typing import (
+    Any,
+    Callable,
+    Iterable,
+    Iterator,
+    Mapping,
+    Tuple,
+)
+import os
+from abc import ABC
+from opentelemetry import trace as trace_api
+from contextlib import contextmanager
+from opentelemetry.util.types import AttributeValue
+from opentelemetry.trace import INVALID_SPAN
+from agentscope.models.response import ModelResponse
+
+from functools import wraps
+
+ENABLE_AGENTSCOPE_INSTRUMENTOR = "ENABLE_AGENTSCOPE_INSTRUMENTOR"
+
+class _WithTracer(ABC):
+    def __init__(self, tracer: trace_api.Tracer, *args: Any, **kwargs: Any) -> None:
+        super().__init__(*args, **kwargs)
+        self._request_attributes_extractor = RequestAttributesExtractor()
+        self._response_attributes_extractor = ResponseAttributesExtractor()
+        self._tracer = tracer
+
+    @contextmanager
+    def _start_as_current_span(
+            self,
+            span_name: str,
+            attributes: Iterable[Tuple[str, AttributeValue]],
+            extra_attributes: Iterable[Tuple[str, AttributeValue]],
+    ) -> Iterator[_WithSpan]:
+        # Because OTEL has a default limit of 128 attributes, we split our attributes into
+        # two tiers, where the addition of "extra_attributes" is deferred until the end
+        # and only after the "attributes" are added.
+        try:
+            span = self._tracer.start_span(name=span_name, attributes=dict(attributes))
+        except Exception:
+            span = INVALID_SPAN
+        with trace_api.use_span(
+                span,
+                end_on_exit=False,
+                record_exception=False,
+                set_status_on_exception=False,
+        ) as span:
+            yield _WithSpan(span=span, extra_attributes=dict(extra_attributes))
+
+class AgentscopeRequestWrapper(_WithTracer):
+
+    def __init__(self, tracer, *args, **kwargs):
+        super().__init__(tracer, *args, **kwargs)
+        self.class_replaced = []
+
+    def _finalize_response(
+            self,
+            response: Any,
+            with_span: _WithSpan,
+    ) -> Any:
+        resp_attr = self._response_attributes_extractor.extract(response)
+        with_span.finish_tracing(
+            status=trace_api.Status(status_code=trace_api.StatusCode.OK),
+            attributes=dict(resp_attr),
+            extra_attributes=dict(resp_attr),
+        )
+        return response
+
+    def _is_enable(self):
+        enable_instrumentor = os.getenv(ENABLE_AGENTSCOPE_INSTRUMENTOR)
```

Comment:
Make sure the environmental variables follow: https://opentelemetry.io/docs/zero-code/python/configuration/